clang/lib/CConv/ConstraintVariables.cpp

```diff
@@ -1647,19 +1645,57 @@ void FunctionVariableConstraint::mergeDeclaration(ConstraintVariable *FromCV) {
   auto RetVar = getOnly(returnVars);
   RetVar->mergeDeclaration(FromRetVar);
   // Transplant params.
-  assert(From->numParams() == numParams());
-  for (unsigned i = 0; i < From->numParams(); i++) {
-    CVarSet &FromP = From->getParamVar(i);
-    CVarSet &P = getParamVar(i);
-    auto FromVar = getOnly(FromP);
-    auto Var = getOnly(P);
-    Var->mergeDeclaration(FromVar);
+  handle_params(From, this,
+                [](ConstraintVariable *F, ConstraintVariable *T)
+                { T->mergeDeclaration(F); });
+}
+
+void FunctionVariableConstraint::handle_params
+(FVConstraint *From, FVConstraint *To,
+ std::function<void(ConstraintVariable*,ConstraintVariable*)> f) {
+  bool fromEmpty = From->numParams() == 0;
+  bool toEmpty = To->numParams() == 0;
+  bool paramsEq = From->numParams() == To->numParams();
+  assert(fromEmpty || toEmpty || paramsEq);
+  if (paramsEq) {
+    for (unsigned i = 0; i < From->numParams(); i++) {
+      CVarSet &FromP = From->getParamVar(i);
+      CVarSet &P = To->getParamVar(i);
+      auto FromVar = getOnly(FromP);
+      auto Var = getOnly(P);
+      f(FromVar, Var);
+    }
+  } else { // Dealing with an untyped prototype
+    FVConstraint *Empty = fromEmpty ? From : To;
+    FVConstraint *Typed = fromEmpty ? To : From;
```

Comment:
What about the case that they are both empty? I think you want to do nothing, then.

Reply:
(True whether they are `f()` style and `f(void)` style.)

Reply:
If they are both empty, the the number of params will be equal and it will be handled in the first block, or am I missing something? Should I be more explicit about that?

Reply:
OK good point.